documentation/puppet/v3/environment_classes.markdown

```diff
@@ -0,0 +1,294 @@
+ 
+### Headers and Caching Behavior
+
+If the `environment-class-cache-enabled` setting in the `jruby-puppet`
+configuration section is set to `true`, the `environment_classes` API can
+make use of caching for the response data.  This can provide a significant
+performance benefit and reduction in the amount of data that needs to be
+provided in a response when the underlying Puppet code on disk remains
+unchanged from one request to the next.  Use of the cache does, however,
+require that cache entries are invalidated after Puppet code has been
+updated.  To avoid having to invalidate cache entries, the
+`environment-class-cache-enabled` setting can be omitted from configuration 
+or explicitly set to `false`.  In this case, manifests will be re-discovered
+from disk and re-parsed for every incoming request.  This can involve
+significantly greater overhead in performance and bandwidth for repeated 
+requests where the underlying Puppet code does not change much.  This 
+approach does, however, ensure that the latest available data is returned 
+for every request made.  The rest of the content in this section pertains to the
+behaviors to keep in mind when the cache setting is enabled.
+
+When the `environment-class-cache-enabled` setting is set to `true`, the 
+response to a query to the `environment_classes`  endpoint includes an HTTP
+[Etag](https://tools.ietf.org/html/rfc7232#section-2.3) header.  The value 
+for the Etag header is a hash which represents the state of the latest class
+information available for the requested environment.  For example:
+
+~~~
+ETag: 31d64b8038258202b4f5eb508d7dab79c46327bb
+~~~
+
+A client may (but is not required to) provide the Etag value back to the server
+in a subsequent `environment_classes` request.  The client would provide the
+tag value as the value for an
+[If-None-Match](https://tools.ietf.org/html/rfc7232#section-3.2)
+HTTP header:
+
+~~~
+If-None-Match: 31d64b8038258202b4f5eb508d7dab79c46327bb
+~~~
+
+If the latest state of code available on the server matches that of the value
+in the `If-None-Match` header, the server will provide an HTTP 304 (Not 
+Modified) response and no corresponding response body.  If the server has later
+code available than what is captured by the `If-None-Match` header value, the
+server will re-parse code from manifest files on disk.  Assuming the resulting
+payload were different than in a previous request, the server would provide a
+different Etag value along new class information in the response payload.
+
+---
+
+After Puppet code (manifests) on disk in an environment are updated, the cache 
+entries that the server holds for class information will need to be cleared. 
+This will allow the server to re-parse the latest manifest code on disk and 
+reflect any changes to the class information in later queries to the 
+environment_classes endpoint.  The following actions will clear cache entries on
+the server:
+
+1) Calling the
```

Comment:
Not a blocker, but ideally these would use the normal markdown ordered list syntax of `1.`.